## Data Cleaning

There are 3 files that I will use for this project. All of the files need to be checked and cleaned.

### 1. Listings

This data is published in: http://insideairbnb.com/get-the-data/

In [1]:
import glob #to return file paths with a specific pattern
import pandas as pd #data processing
import os #interaction with operating system
import numpy as np #working with arrays

In [2]:
#change directory
os.chdir('C:\\Users\\User\\Desktop\\airbnb_data')

In [3]:
#input data
dflisting = pd.read_csv('listings.csv')
#have a look at the data
dflisting.head(5)

id                                      listing_url  \
0  607014960957869901  https://www.airbnb.com/rooms/607014960957869901   
1             1040492             https://www.airbnb.com/rooms/1040492   
2               15420               https://www.airbnb.com/rooms/15420   
3               24288               https://www.airbnb.com/rooms/24288   
4               38628               https://www.airbnb.com/rooms/38628   

        scrape_id last_scraped  \
0  20220610041446   2022-06-10   
1  20220610041446   2022-06-10   
2  20220610041446   2022-06-10   
3  20220610041446   2022-06-10   
4  20220610041446   2022-06-10   

                                                name  \
0           Beautiful 3 bed family home with hot tub   
1            Edinburgh Waterfront Apartments (green)   
2                  Georgian Boutique Apt City Centre   
3  Cool central Loft, sleeps 4, 2 double bed+en-s...   
4                              Edinburgh Holiday Let   

                                         description  \
0  Your family will be close to everything when y...   
1  <b>The space</b><br />Set in Edinburgh’s fashi...   
2  Stunning, impeccably refurbished spacious grou...   
3  Upper level of duplex. Boho rustic-chic former...   
4  Brunstane - Daiches Braes (close to Portobello...   

                               neighborhood_overview  \
0                                                NaN   
1                                                NaN   
2  The neighbourhood is in the historic New Town,...   
3  It's all in the mix: Culture-museums and galle...   
4                  Quiet and easy access to outside.   

                                         picture_url    host_id  \
0  https://a0.muscache.com/pictures/22ba6932-11e6...  454666979   
1  https://a0.muscache.com/pictures/18975884/b655...    1297674   
2  https://a0.muscache.com/pictures/cf69631f-4194...      60423   
3  https://a0.muscache.com/pictures/3460007/88731...      46498   
4  https://a0.muscache.com/pictures/d9885120-178e...     165635   

                                      host_url  ...  \
0  https://www.airbnb.com/users/show/454666979  ...   
1    https://www.airbnb.com/users/show/1297674  ...   
2      https://www.airbnb.com/users/show/60423  ...   
3      https://www.airbnb.com/users/show/46498  ...   
4     https://www.airbnb.com/users/show/165635  ...   

  review_scores_communication review_scores_location review_scores_value  \
0                         NaN                    NaN                 NaN   
1                        4.90                   4.64                4.70   
2                        4.98                   4.98                4.90   
3                        4.89                   4.85                4.60   
4                        4.80                   4.73                4.75   

  license instant_bookable calculated_host_listings_count  \
0     NaN                t                              1   
1     NaN                t                              3   
2     NaN                f                              1   
3     NaN                t                              1   
4     NaN                f                              2   

  calculated_host_listings_count_entire_homes  \
0                                           1   
1                                           3   
2                                           1   
3                                           1   
4                                           2   

  calculated_host_listings_count_private_rooms  \
0                                            0   
1                                            0   
2                                            0   
3                                            0   
4                                            0   

  calculated_host_listings_count_shared_rooms reviews_per_month  
0                                           0               NaN  
1                                           0              0.58  
2     

In [4]:
#check the number of rows and columns
dflisting.shape

(6767, 74)

In [5]:
#check the data type of each attribute
dflisting.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6767 entries, 0 to 6766
Data columns (total 74 columns):
 #   Column                                        Non-Null Count  Dtype  
---  ------                                        --------------  -----  
 0   id                                            6767 non-null   int64  
 1   listing_url                                   6767 non-null   object 
 2   scrape_id                                     6767 non-null   int64  
 3   last_scraped                                  6767 non-null   object 
 4   name                                          6767 non-null   object 
 5   description                                   6690 non-null   object 
 6   neighborhood_overview                         4976 non-null   object 
 7   picture_url                                   6767 non-null   object 
 8   host_id                                       6767 non-null   int64  
 9   host_url                                      6767 non-null   o

In [6]:
#check lost data
dflisting.isnull().sum()

id                                                 0
listing_url                                        0
scrape_id                                          0
last_scraped                                       0
name                                               0
                                                ... 
calculated_host_listings_count                     0
calculated_host_listings_count_entire_homes        0
calculated_host_listings_count_private_rooms       0
calculated_host_listings_count_shared_rooms        0
reviews_per_month                               1007
Length: 74, dtype: int64

There are 74 columns in this dataset. I will not need all of them, thus I am only keeping the ones I consider important for my analysis and those which are not null.

In [7]:
#keep only certain columns
dflistingclean = dflisting[['id', 'listing_url', 'name', 'host_response_time',
                'host_total_listings_count', 'neighbourhood', 'neighbourhood_cleansed', 
                'latitude', 'longitude', 'property_type', 'room_type', 'accommodates', 
                 'bedrooms', 'beds', 'amenities', 'price', 'minimum_nights',
                'maximum_nights', 'availability_30', 'reviews_per_month']]

In [8]:
dflistingclean.head(5)

id                                      listing_url  \
0  607014960957869901  https://www.airbnb.com/rooms/607014960957869901   
1             1040492             https://www.airbnb.com/rooms/1040492   
2               15420               https://www.airbnb.com/rooms/15420   
3               24288               https://www.airbnb.com/rooms/24288   
4               38628               https://www.airbnb.com/rooms/38628   

                                                name  host_response_time  \
0           Beautiful 3 bed family home with hot tub      within an hour   
1            Edinburgh Waterfront Apartments (green)  within a few hours   
2                  Georgian Boutique Apt City Centre      within an hour   
3  Cool central Loft, sleeps 4, 2 double bed+en-s...      within an hour   
4                              Edinburgh Holiday Let        within a day   

   host_total_listings_count                                 neighbourhood  \
0                          0                                           NaN   
1                          3                                           NaN   
2                          3  Edinburgh, City of Edinburgh, United Kingdom   
3                          1            Edinburgh, EH8 9JW, United Kingdom   
4                          2  Edinburgh, City of Edinburgh, United Kingdom   

                      neighbourhood_cleansed  latitude  longitude  \
0          Jewel, Brunstane and Newcraighall  55.93566   -3.07882   
1            Western Harbour and Leith Docks  55.98879   -3.18733   
2  Old Town, Princes Street and Leith Street  55.95759   -3.18805   
3       Canongate, Southside and Dumbiedykes  55.94383   -3.18445   
4                                      Joppa  55.94215   -3.09640   

        property_type        room_type  accommodates  bedrooms  beds  \
0         Entire home  Entire home/apt             6       3.0   3.0   
1  Entire rental unit  Entire home/apt             6       2.0   4.0   
2  Entire rental unit  Entire home/apt             2       1.0   1.0   
3         Entire loft  Entire home/apt             4       2.0   2.0   
4  Entire rental unit  Entire home/apt             2       1.0   2.0   

                                           amenities    price  minimum_nights  \
0  ["BBQ grill", "Free dryer \u2013 In building",...  $300.00               5   
1  ["Waterfront", "Pack \u2019n play/Travel crib"...  $205.00               2   
2  ["Pack \u2019n play/Travel crib", "Refrigerato...  $110.00               3   
3  ["Refrigerator", "Coffee maker", "Smoke alarm"...   $95.00               3   
4  ["Waterfront", "Refrigerator", "Coffee maker",...   $51.00               4   

   maximum_nights  reviews_per_month  
0             365                NaN  
1             365               0.58  
2              30               2.95  
3             365               1.60  
4             120               0.64

In [9]:
#check shape of dataframe columns 
dflistingclean.columns.shape

(19,)

In [10]:
#fill empty rows with 0
dflistingclean['reviews_per_month'].fillna(0, inplace=True)

C:\Users\User\anaconda3\lib\site-packages\pandas\core\series.py:4463: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().fillna(


In [11]:
#save dataframe to csv file
dflistingclean.to_csv('listing_cleansed.csv', index=False)

### 2. Calendar 

This data is published in: http://insideairbnb.com/get-the-data/

Read the four calendar files

In [12]:
#change directory
os.chdir('C:\\Users\\User\\Desktop\\airbnb_data\\calendar_files')

In [13]:
#open first calendar file and check size
calendar202109 = pd.read_csv("calendar-202109.csv") 
calendar202109.size

15807785

In [14]:
#open second calendar file and check size
calendar202112 = pd.read_csv("calendar-202112.csv")
calendar202112.size

15621270

In [15]:
#open third calendar file and check size
calendar202203 = pd.read_csv("calendar-202203.csv")
calendar202203.size

15741292

In [16]:
#open fourth calendar file and check size
calendar202206 = pd.read_csv("calendar-202206.csv")
calendar202206.size

17289685

At this point I want to check whether calendar files are within specific date range

In [17]:
#change datatype of date column and check dates
calendar202109['date']= calendar202109['date'].astype('datetime64[ns]')
a = calendar202109['date'].dt.date.unique()
((a > pd.Timestamp('2021-09-01')) & (a < pd.Timestamp('2022-09-30'))).all()

True

In [18]:
#change datatype of date column and check dates
calendar202112['date']= calendar202112['date'].astype('datetime64[ns]')
b = calendar202112['date'].dt.date.unique()
((b > pd.Timestamp('2021-12-01')) & (b < pd.Timestamp('2022-12-31'))).all()

True

In [19]:
#change datatype of date column and check dates
calendar202203['date']= calendar202203['date'].astype('datetime64[ns]')
c = calendar202203['date'].dt.date.unique()
((c > pd.Timestamp('2022-03-01')) & (c < pd.Timestamp('2023-03-31'))).all()

True

In [20]:
#change datatype of date column and check dates
calendar202206['date']= calendar202206['date'].astype('datetime64[ns]')
d = calendar202206['date'].dt.date.unique()
((d > pd.Timestamp('2022-06-01')) & (d < pd.Timestamp('2023-06-30'))).all()

True

After checking the data, I will compile all the files

In [21]:
#setting the path for joining multiple files
files = os.path.join( "*.csv")

#return list of merged files
files = glob.glob(files)

#joining files with concat and read_csv
dfcalendar = pd.concat(map(pd.read_csv, files), ignore_index=True)

In [22]:
#check columns and rows
dfcalendar.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9208576 entries, 0 to 9208575
Data columns (total 7 columns):
 #   Column          Dtype  
---  ------          -----  
 0   listing_id      int64  
 1   date            object 
 2   available       object 
 3   price           object 
 4   adjusted_price  object 
 5   minimum_nights  float64
 6   maximum_nights  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 491.8+ MB


In [23]:
#view first 5 rows of dataframe 
dfcalendar.head(5)

listing_id        date available    price adjusted_price  minimum_nights  \
0       15420  2021-09-16         f  $100.00        $100.00             3.0   
1       93904  2021-09-16         f   $55.00         $55.00             1.0   
2       93904  2021-09-17         f   $55.00         $55.00             1.0   
3       93904  2021-09-18         f   $55.00         $55.00             1.0   
4       93904  2021-09-19         f   $55.00         $55.00             1.0   

   maximum_nights  
0            30.0  
1            31.0  
2            31.0  
3            31.0  
4            31.0

In [24]:
#check lost data
dfcalendar.isnull().sum()

listing_id        0
date              0
available         0
price             0
adjusted_price    0
minimum_nights    2
maximum_nights    2
dtype: int64

There are 2 rows in two columns which are null

In [25]:
#check which rows are null
dfcalendar[dfcalendar['minimum_nights'].isna()]

listing_id        date available    price adjusted_price  \
2797148    14041276  2021-12-12         f  $139.00        $139.00   
2872109    16342332  2021-12-12         f   $66.00         $66.00   

         minimum_nights  maximum_nights  
2797148             NaN             NaN  
2872109             NaN             NaN

In [26]:
#find first the row
dfcalendar.loc[(dfcalendar['listing_id'] == 14041276) & (dfcalendar['date'] == '2021-12-12')]

listing_id        date available    price adjusted_price  \
530314     14041276  2021-12-12         t  $119.00        $119.00   
2797148    14041276  2021-12-12         f  $139.00        $139.00   

         minimum_nights  maximum_nights  
530314              2.0          1125.0  
2797148             NaN             NaN

In [27]:
#find second the row
dfcalendar.loc[(dfcalendar['listing_id'] == 16342332) & (dfcalendar['date'] == '2021-12-12')]

listing_id        date available   price adjusted_price  \
623483     16342332  2021-12-12         t  $79.00         $79.00   
2872109    16342332  2021-12-12         f  $66.00         $66.00   

         minimum_nights  maximum_nights  
623483              2.0          1125.0  
2872109             NaN             NaN

There are two rows with null values that are duplicates

In [28]:
#drop null rows
dfcalendar = dfcalendar.dropna()

In [29]:
#drop adjusted_price column
dfcalendar = dfcalendar.drop(['adjusted_price'], axis=1) 

In [30]:
#change datatypes of columns
dfcalendar['listing_id'] = dfcalendar['listing_id'].astype(int)
dfcalendar['minimum_nights'] = dfcalendar['minimum_nights'].astype(int)
dfcalendar['maximum_nights'] = dfcalendar['maximum_nights'].astype(int)
dfcalendar['date']= dfcalendar['date'].astype('datetime64[ns]')
dfcalendar.head(5)

listing_id       date available    price  minimum_nights  maximum_nights
0       15420 2021-09-16         f  $100.00               3              30
1       93904 2021-09-16         f   $55.00               1              31
2       93904 2021-09-17         f   $55.00               1              31
3       93904 2021-09-18         f   $55.00               1              31
4       93904 2021-09-19         f   $55.00               1              31

In [31]:
#remove duplicates based on two columns
dfcalendar = dfcalendar.drop_duplicates(
  subset = ['listing_id', 'date'],
  keep = 'last').reset_index(drop = True)

In [32]:
 #check format of price numbers
dfcalendar['price'].unique()   

array(['$100.00', '$55.00', '$50.00', ..., '$1,950.00', '$20,000.00',
       '$6,784.00'], dtype=object)

Price column has dollar symbols and commas

In [33]:
#remove dollar symbol and comma from rows of price column
dfcalendar['price'] = dfcalendar['price'].replace({'\$': '', ',': ''}, regex=True).astype(float)

In [34]:
#check format of price numbers again
print(dfcalendar['price'].unique())

[  100.    55.    50. ...  1950. 20000.  6784.]


In [35]:
#number of accommodations based on ids
dfcalendar['listing_id'].nunique()

8415

The dataframe with all the calendar data contains 8415 unique listings

In [36]:
#save dataframe to csv file
dfcalendar.to_csv('all_calendar_data.csv', index=False)

### 3. Broader areas

This data is published in: https://statistics.gov.scot/atlas/resource?uri=http%3A%2F%2Fstatistics.gov.scot%2Fid%2Fstatistical-geography%2FS12000036

Concatenate the neighbourhood files 

In [37]:
#change directory one last time
os.chdir('C:\\Users\\User\\Desktop\\airbnb_data\\broader_areas')

In [38]:
#setting the path for joining multiple files
files = os.path.join( "*.csv")

#list of merged files returned
files = glob.glob(files)

#joining files with concat and read_csv
dfareas = pd.concat(map(pd.read_csv, files), ignore_index=True)

In [39]:
#view first 5 rows of dataframe
dfareas.head(5)

Feature Identifier  \
0  http://statistics.gov.scot/id/statistical-geog...   
1  http://statistics.gov.scot/id/statistical-geog...   
2  http://statistics.gov.scot/id/statistical-geog...   
3  http://statistics.gov.scot/id/statistical-geog...   
4  http://statistics.gov.scot/id/statistical-geog...   

                     Feature_Name Electoral_Ward  
0  Granton West and Salvesen - 05         Almond  
1                  Muirhouse - 01         Almond  
2                  Muirhouse - 02         Almond  
3                  Muirhouse - 03         Almond  
4                  Muirhouse - 04         Almond

In [40]:
#drop first column and view dataframe again
dfareas=dfareas.iloc[: , 1:]
dfareas.head(5)

Feature_Name Electoral_Ward
0  Granton West and Salvesen - 05         Almond
1                  Muirhouse - 01         Almond
2                  Muirhouse - 02         Almond
3                  Muirhouse - 03         Almond
4                  Muirhouse - 04         Almond

In [41]:
#split and remove the strings after delimiter and view
dfareas['Feature_Name'] = dfareas['Feature_Name'].str.split(' -').str[0]
dfareas.head(5)

Feature_Name Electoral_Ward
0  Granton West and Salvesen         Almond
1                  Muirhouse         Almond
2                  Muirhouse         Almond
3                  Muirhouse         Almond
4                  Muirhouse         Almond

In [42]:
#remove duplicates based on two columns
dfareas = dfareas.drop_duplicates(subset=['Feature_Name'])
dfareas.columns = ['neighbourhood_cleansed', 'broader_area']
print(dfareas)

                      neighbourhood_cleansed         broader_area
0                  Granton West and Salvesen               Almond
1                                  Muirhouse               Almond
7          Silverknowes and Davidson's Mains               Almond
13                                   Cramond               Almond
16          Barnton, Cammo and Cramond South               Almond
..                                       ...                  ...
547                              Gorgie East     Sighthill/Gorgie
559  Blackford, West Mains and Mayfield Road  Southside/Newington
567                             Prestonfield  Southside/Newington
571              Newington and Dalkeith Road  Southside/Newington
576                               The Grange  Southside/Newington

[111 rows x 2 columns]


In [43]:
#save dataframe to csv file
dfareas.to_csv('broader_areas.csv', index=False)